### ISSUE TITLE
Write error when printing stdin

### ISSUE THREAD

**Issue by kirbyfan64** [kevinlawler/kona#366]
Example program:

```
`0:0:`"/dev/stdin"
```

Output with K2:

```
ryan@DevPC-LX:~/golf/lcs$ echo | k2 bug.k


ryan@DevPC-LX:~/golf/lcs$ 
```

Output with Kona:

```
ryan@DevPC-LX:~/golf/lcs$ echo | k bug.k

write error
`0:0:`"/dev/stdin"
at execution instance 1 of "0:"
>  ryan@DevPC-LX:~/golf/lcs$ 
```

Dang, I feel like a bug magnet sometimes... :O

**Comment by tavmem** [comment]
A more fundamental difference:
In both k2.8 and Kona, you can get exactly the same resuts in the following:
k2.8:
```
   0:`"/dev/stdin"
a
b
(,"a"
 ,"b")
```
Kona:
```
   0:"/dev/stdin"
a
b
(,"a"
 ,"b")
```
However, in k2.8 you have about 1 second to enter the "a", and about 1 second to enter the "b", otherwise k2.8 no longer waits for input, prints the prompt, and returns to normal input mode.

Kona, on the other hand, waits indefinitely for you to input the "a" and again indefinitely for the "b", and will continue to wait until you enter Cntrl-d to terminate input, at which time the prompt is printed and Kona returns to normal input mode.

For Linux, Kona uses getdelim() which seems to block indefintely.  This would seem to mean that k2.8 does not use getdelim() here.

**Comment by tavmem** [comment]
See
http://stackoverflow.com/questions/717572/how-do-you-do-non-blocking-console-i-o-on-linux-in-c

**Comment by tavmem** [comment]
On the other hand, in this case:
```
  0:`
abc
"abc"
```
both k2.8 and Kona wait indefinitely for the input "abc", then return immediately.

**Comment by tavmem** [comment]
Actually, in the following case (in k2.8):
```
   x:0:`"/dev/stdin"
a
a
a
a
  a
value error
a
^
parse error
  x
(,"a"
 ,"a"
 ,"a"
 ,"a")
```
you have 1 or 2 seconds to enter all input.  I was able to enter "a" four times. The standard input mode received the fifth "a", yielding "value error" and "parse error".

**Comment by kirbyfan64** [comment]
I think what k2 does has something to do with this: http://stackoverflow.com/questions/5616092/non-blocking-call-for-reading-descriptor. It probably just loops for up to a certain time limit while there is no input.

**Comment by tavmem** [comment]
It's possible ....

I think that I prefer the following approach:
http://stackoverflow.com/a/30548692/3873997

**Comment by tavmem** [comment]
Using a combination of fcntl() and read() works in that you can limit the time during which input is accepted before returning to standard input mode.
However, read() only accepts one line of input, so you can only enter one of the "a"s.

**Next step:** Try some other variations.  Ncurses would probably work, but linking to the ncurses library may add significantly to the footprint of the Kona executable.

**Comment by tavmem** [comment]
Commit 6b7adfa65744707c0ce39eb1fcd18216417b54be
fixes the inconsistency regarding 0:`"/dev/stdin"

**Next step:** Revist the "write error" for this issue.

**Comment by tavmem** [comment]
A related problem in Kona with a different manifestation ("type error" instead of "write error"):
```
tom@xps ~/kona $ cat bug.k
`0:0:`"/dev/stdin"

tom@xps ~/kona $ rlwrap ./k
K Console - Enter \ for help

  \l bug.k
type error
> 
```

**Comment by bakul** [comment]
On Wed, 15 Jul 2015 06:36:41 PDT Tom Szczesny <notifications@github.com> wrote:
> Commit 6b7adfa65744707c0ce39eb1fcd18216417b54be
> fixes the inconsistency regarding 0:`"/dev/stdin"

This fix doesnt seem right. With sleep(1), even if the data is
ready the program will take an extra second.

I haven't examined the code but in general, if you need a
timeout, On unix you can use select() or poll() with a timeout
so that it will return as soon as data is available or it will
timeout.

The orignal bug doesn't seem to be related to timeout at all.
To recap:

    cat > echo.k <<'EOF'
    #!env kona
    `0:0:`"/dev/stdin"
    EOF
    chmod +x echo.k
    date|echo.k
    Wed Jul 15 13:10:10 PDT 2015
    echo|echo.k

    write error
    `0:0:`"/dev/stdin"
    at execution instance 1 of "0:"
    >

This seems to be related to not dealing with a null line
correctly. May be getdelim() is not the right thing to use?
grepping for getdelim I see

src/0.c:88:    b=getdelim_(&v,(size_t * __restrict__)&s,EOF,stdin);

This seems wrong too since EOF is not a delimiter. The reason
getdelim() was introduced in the first place was because
people may have already been using getline() in their code.
But for new code getline() shoud be used.

getline() is provided on unix platforms (but you may need to
specify a platform specific #define). Not sure what is
available on windows but other platforms should not pay the
cost for windows' sins. Ideally all machine dependencies
should be confined to one files and the rest of the code be
devoid of any machine dependent #ifdefs.

Finally the timeout behavior of k-2.8 was fixed in k-3.2 so it
was clearly a bug! k-3.2 waits until the user types ^D to end
the input. The only different between ` and `"/dev/stdin" is
that the former accepts a single line, while the latter
accepts all input until the end of file.

**Comment by tavmem** [comment]
What "doesn't seem right" to me is the k2.8 implementation of 0:`"/dev/stdin"
It is not so much that it provides a "timeout", but that it imposes a very small window of time for data entry.
Using k2.8 try to enter the string 135792468:
```
  x:0:`"/dev/stdin"
  135792468
135792468
  x
()
```
You don't have time to do it ... which does not make any sense.

I much prefer the implemetation that Kona had before this issue was opened.
```
  x:0:`"/dev/stdin"
abd
3333333
trr
yyx
("abd"
 "3333333"
 "trr"
 "yyx")
```
You have as much time as you want to enter data.
You can enter as many lines as you wish.
You terminate entry with Cntl-d.

I never claimed that the original bug was related to this k2.8/Kona inconsitency.
The original bug was
```
cat bug.k
`0:0:`"/dev/stdin"
ryan@DevPC-LX:~/golf/lcs$ echo | k bug.k
write error
```
In looking for simpler syntactic occurrences, I noticed that
```
  \l bug.k
type error
```
and that
```
  0:`"/dev/stdin"
```
behaves completely differently in k2.8 than in Kona.

I was operating under the assumption that Arthur had some reason for the k2.8 implementation of
```
  0:`"/dev/stdin"
```
If there is no justifiable reason, we should go back to the implementation that existed in Kona.
